File: Data_Structures_and_Algorithms/w4_sorting_and_searching/binary_search.py
# binary search algorithm in Python

from ast import And


def binary_search(list, value):
	first = 0		 	# first index of list
	last = len(list)-1  # last index of list
	
	while(last >= first):
		middle = (first+last)//2

	# this doesn't work right cuz overflows the list
	# while(True):
	# 	middle = (first+last)//2	
		# if (first == middle or middle == last) and list[middle] != value:
		# 	return -1

		if list[middle] == value:
			return middle
		elif value < list[middle]:
			last = middle-1
		elif value > list[middle]:
			first = middle+1

	return -1

list = [10, 14, 19, 26, 27, 30, 31, 33, 35, 42, 44]

test_list = [2, 10, 19, 30, 42, 44, 57];

for value in test_list:
	result = binary_search(list, value)

	if result != -1:
		print(f"Index of value {value} is {result}")
	else:
		print(f"{value} is not in the list.")

File: Data_Structures_and_Algorithms/w4_sorting_and_searching/binary_search.c
// binary search in C

#include <stdio.h>

int binary_search(int list[], int size_of_list, int value){
	int first = 0;
	int last = size_of_list-1;
	int middle;

	while(last >= first){
		middle = (first+last)/2;

	// this doesn't work right cuz overflows the list
	// while(){
	// 	middle = (first+last)/2;
		// if(first==middle && last==middle && list[middle]!=value)
		// 	return -1;

		if (value == list[middle])
			return middle;

		else if(value < list[middle])
			last = middle-1;

		else if(value > list[middle])
			first = middle+1;

		//printf("numbers: %d %d %d\n", first, last, middle);
	}
	return -1;
}


int main(){

	int list[] = {10, 14, 19, 26, 27, 30, 31, 33, 35, 42, 44};
	int size_of_list = sizeof(list)/sizeof(list[0]);
	int value = 14;

	int test_list[] = {2, 10, 19, 30, 42, 44, 57};
	int size_of_test_list = sizeof(test_list)/sizeof(test_list[0]);

	for(int i=0; i<size_of_test_list; i++){
		value = test_list[i];

		int result = binary_search(list, size_of_list, value);

		if(result >= 0)
			printf("The index of %d is %d.\n", value, result);
		else if(result==-1)
			printf("%d is not found.\n", value);
	}
	return 0;
}